crates/pessimistic-proof-utils/src/smt.rs

```diff
@@ -0,0 +1,559 @@
+                Err(SmtError::KeyAlreadyPresent)
+            } else {
+                Ok(value)
+            };
+        }
+        let node = self.tree.get(&hash);
+        let mut node = node.copied().unwrap_or(Node {
+            left: self.empty_hash_at_height[DEPTH - depth - 1],
+            right: self.empty_hash_at_height[DEPTH - depth - 1],
+        });
+        let child_hash = if bits[depth] {
+            self.insert_helper(node.right, depth + 1, bits, value)
+        } else {
+            self.insert_helper(node.left, depth + 1, bits, value)
+        }?;
+        if bits[depth] {
+            node.right = child_hash;
+        } else {
+            node.left = child_hash;
+        }
+
+        let new_hash = node.hash();
+        self.tree.insert(new_hash, node);
+
+        Ok(new_hash)
+    }
+
+    pub fn insert<K>(&mut self, key: K, value: H::Digest) -> Result<(), SmtError>
+    where
+        K: ToBits<DEPTH>,
+    {
+        let new_root = self.insert_helper(self.root, 0, &key.to_bits(), value)?;
+        self.root = new_root;
+
+        Ok(())
+    }
+
+    pub fn get_inclusion_proof<K>(&self, key: K) -> Result<SmtMerkleProof<H, DEPTH>, SmtError>
+    where
+        K: ToBits<DEPTH>,
+    {
+        let mut siblings = [self.empty_hash_at_height[0]; DEPTH];
+        let mut hash = self.root;
+        let bits = key.to_bits();
+        for i in 0..DEPTH {
+            let node = self.tree.get(&hash).ok_or(SmtError::KeyNotPresent)?;
+            siblings[DEPTH - i - 1] = if bits[i] { node.left } else { node.right };
+            hash = if bits[i] { node.right } else { node.left };
+        }
+        if hash == self.empty_hash_at_height[0] {
+            return Err(SmtError::KeyNotPresent);
+        }
+
+        Ok(SmtMerkleProof { siblings })
+    }
+
+    pub fn get_non_inclusion_proof<K>(
+        &self,
+        key: K,
+    ) -> Result<SmtNonInclusionProof<H, DEPTH>, SmtError>
+    where
+        K: ToBits<DEPTH>,
+    {
+        let mut siblings = vec![];
+        let mut hash = self.root;
+        let bits = key.to_bits();
+        for i in 0..DEPTH {
+            if self.empty_hash_at_height.contains(&hash) {
+                return Ok(SmtNonInclusionProof { siblings });
+            }
+            let node = self.tree.get(&hash);
+            let node = match node {
+                Some(node) => node,
+                None => {
+                    debug_assert!(
+                        hash == H::merge(
+                            &self.empty_hash_at_height[DEPTH - i - 1],
+                            &self.empty_hash_at_height[DEPTH - i - 1]
+                        ),
+                        "The SMT is messed up"
+                    );
+                    return Ok(SmtNonInclusionProof { siblings });
+                }
+            };
+            siblings.push(if bits[i] { node.left } else { node.right });
+            hash = if bits[i] { node.right } else { node.left };
```

Comment:
You can reduce the boilerplate to this, I guess: 
```suggestion
            hash = if bits[i] { node.right } else { node.left };
            siblings.push(hash);
            
```